beacon/blockchain/process_proposal.go

```diff
@@ -280,10 +270,19 @@ func (s *Service) VerifyIncomingBlock(
 		return ErrUnexpectedBlockSlot
 	}
 
+	var preFetchFailureData *preFetchedBuildData
+	if s.shouldBuildOptimisticPayloads() {
+		copiedState := state.Copy(ctx)
+		preFetchFailureData, err = s.preFetchBuildDataForRejection(copiedState, consensusTime)
+		if err != nil {
+			return fmt.Errorf("failed preFetching data for rejection: %w", err)
+		}
+	}
+
 	// Verify the state root of the incoming block.
 	err = s.verifyStateRoot(
 		ctx,
-		postState,
+		state,
```

Comment:
q: why shouldn't this be a copy of the state? i.e. if we end up rejecting the block here (verifyStateRoot returns non-nil err) then we dont want to persist any changes on state itself?

Reply:
Or does writing to this ctx not matter as it never gets persisted anyhow?

Reply:
The important point is:
- changes carried out on `state` will be visible from `copiedState` but not viceversa.

So, as you said, before this PR we used copiedState to make sure that whatever partial change an invalid block would carried out won't pollute `state` which was used to optimistically request a new block.

With my changes however, this does not matter as we pre-fetch all the relevant data.
Also in upcoming state caching we do need to carry out the block changes on `state` in order to cache it and not on `copiedState`. So I took advantage of this PR to do that.

This is a subtle but important point. LMK if this makes sense to you